Allow local build of remote_server dev to be deployed to different linux than local

gwen-lg opened the pull request:
setup local build of `remote_server` to not depend of the local linux libraries by :
- enable `vendored-libgit2` feature of git2
- setup target triple to `unknown-linux-musl` (mirror bundle-linux script)
- add flag ` -C target-feature=+crt-static` in `RUSTFLAGS` env var (mirror bundle-linux script)

Bonus:
Add an option to setup mold as linker of local build.

Closes #33341

Release Notes:

 - N/A

gwen-lg committed:
add option to use mold for dev build of remote_server

To use it, add `mold` in ZED_BUILD_REMOTE_SERVER value.

gwen-lg committed:
only replace early `return` in `if` by use of `else`

this allow better code readablity and prepare for other code cleaning

gwen-lg committed:
enable feature `vendored-libgit2` by default for remote_server

this avoid to have zed remote_server local build depend on local version
of libgit2 and failed to start on target linux if libgit2 is not installed
or not at the same version.

gwen-lg committed:
deduplicate Compressing binary task in remote_server build code

compression management is now only at the end of the function `build_local`

This fix the `nocompress` functionality handle when cross compiling
but not when not cross compiling .

gwen-lg committed:
inline platform triple fn in build_local

... as the choice of triple is related to what is needed in build_local.

gwen-lg commented:
@SomeoneToIgnore : there is something to do to have approval for awaiting workflow ?

SomeoneToIgnore commented on crates/remote/src/ssh_session.rs:
```diff
@@ -1999,58 +2000,31 @@ impl SshRemoteConnection {
-            delegate.set_status(
-                Some(&format!(
-                    "Building remote server binary from source for {} with Docker",
-                    &triple
```
It's not bad to mention that `triple` is it? 
Any reason to omit?

gwen-lg replied:
I'm not sure to understand your point.
I don't have remove this, but I change the scope (so indentation changed) and the diff is a little lost.

Why do you think about split the build_local function in several smaller functions before ? This can help diff reading.

SomeoneToIgnore replied:
Just, can we put that `tripe` back in the status output?
It's removed in the PR for some reason.

gwen-lg replied:
It's not removed, the leading space change, so the diff mark it as added + removed. You can verify with a search.

gwen-lg replied:
Are you interested in me splitting it into smaller commits to make the change more readable?

SomeoneToIgnore commented on crates/remote/src/ssh_session.rs:
```diff
@@ -2062,69 +2036,111 @@ impl SshRemoteConnection {
-            let which = cx
-                .background_spawn(async move { which::which("zig") })
-                .await;
+            smol::fs::create_dir_all("target/remote_server").await?;
```
That's what `cargo` usually does, why should we help it?

SomeoneToIgnore replied:
Still here, not needed, really.

gwen-lg commented on crates/remote/src/ssh_session.rs:
```diff
@@ -2062,69 +2036,111 @@ impl SshRemoteConnection {
+                .await?;
+            }
+
+            format!("target/remote_server/{triple}/debug/remote_server")
```
Ok, I'm going to take advantage of this PR to make a commit to correct this point.

gwen-lg committed:
use Path instead of String in build_local

also deduplicate remote_server binary path generation

gwen-lg committed:
build remote_server with static musl

This mirror the configuration of release (stable build).
It's allow to deploy the remote_server on different linux than the dev plateform.
Fix issue #33341

gwen-lg commented:
@SomeoneToIgnore I have updated the PR according to your review (if I haven't forgotten anything)

SomeoneToIgnore reviewed (commented):
Nice, thank you for fixing the paths too.

Let's simplify the Cargo.toml -related changes: besides the formatting, it should be able to add just one line with the right features, I think.
Let me know if that does not work for some reason and we can move on.

gwen-lg committed:
remove not needed create_dir_all for target

cargo does

SomeoneToIgnore reviewed (approved):
Great, thank you a lot for bearing through this all.

gwen-lg commented:
Little update to fix build on `windows`.
